charms/metallb-speaker/src/utils.py

```diff
@@ -0,0 +1,224 @@
+"""Kubernetes utils library."""
+
+import logging
+import os
+import random
+import string
+
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+
+logger = logging.getLogger(__name__)
+
+
+def create_pod_security_policy_with_api(namespace):
+    """Create pod security policy."""
+    # Using the API because of LP:1886694
+    logging.info('Creating pod security policy with K8s API')
+    _load_kube_config()
+
+    metadata = client.V1ObjectMeta(
+        namespace=namespace,
+        name='speaker',
+        labels={'app': 'metallb'}
+    )
+    policy_spec = client.PolicyV1beta1PodSecurityPolicySpec(
+        allow_privilege_escalation=False,
+        allowed_capabilities=[
+            'NET_ADMIN',
+            'NET_RAW',
+            'SYS_ADMIN',
+        ],
+        default_allow_privilege_escalation=False,
+        fs_group=client.PolicyV1beta1FSGroupStrategyOptions(
+            rule='RunAsAny'
+        ),
+        host_ipc=False,
+        host_network=True,
+        host_pid=False,
+        host_ports=[client.PolicyV1beta1HostPortRange(
+            max=7472,
+            min=7472,
+        )],
+        privileged=True,
+        read_only_root_filesystem=True,
+        required_drop_capabilities=['ALL'],
+        run_as_user=client.PolicyV1beta1RunAsUserStrategyOptions(
+            rule='RunAsAny'
+        ),
+        se_linux=client.PolicyV1beta1SELinuxStrategyOptions(
+            rule='RunAsAny',
+        ),
+        supplemental_groups=client.PolicyV1beta1SupplementalGroupsStrategyOptions(
+            rule='RunAsAny'
+        ),
+        volumes=['configMap', 'secret', 'emptyDir'],
+    )
+
+    body = client.PolicyV1beta1PodSecurityPolicy(metadata=metadata, spec=policy_spec)
+
+    with client.ApiClient() as api_client:
+        api_instance = client.PolicyV1beta1Api(api_client)
+        try:
+            api_instance.create_pod_security_policy(body, pretty=True)
+            return True
+        except ApiException as err:
+            logging.exception("Exception when calling PolicyV1beta1Api"
+                              "->create_pod_security_policy.")
+            if err.status != 409:
+                # ignoring 409 (AlreadyExists) errors
+                return False
+            else:
+                return True
+
+
+def delete_pod_security_policy_with_api(name):
```

Comment:
I don't know how much duplication there is between these, but if there's enough that you'd like to have some shared code, you can use a relative path in the charm's `requirements.txt`, e.g.:

```
ops
kubernetes
../../../shared/utils/
```

You do need to have a minimal `setup.py` along with the code file (e.g. `utils.py`), something like:

```python
from setuptools import setup

setup(
    name="utils",
    py_modules=["utils"],
)
```

Reply:
As discussed in chat, this isn't really worth spending much effort on unless the shared code gets significantly large, and at that point it might even be worth breaking it out into a separate library all together.  So entirely up to you if you want to even bother with this suggestion; some others might even find it messy or prone to issues (like the caching issue I mentioned in chat).

Reply:
I'm expecting the use of the K8s API to eventually disappear in this code, since it is used only when the juju pod_spec does not support (yet) creating specific resources, and we are tracking the each bug related to this. With that in mind, it might be better to just leave the code as-is, as it should not grow more than its actual state.